File: VB/WindowsFormsSample/Form1.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports DevExpress.Xpo
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Base
Imports DevExpress.Data.Filtering
Imports DevExpress.XtraGrid.Columns
Imports System.Reflection
Imports System.Threading

Namespace WindowsFormsSample

    Public Partial Class Form1
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            For i As Integer = 0 To 5 - 1
                Dim parent As Parent = New Parent()
                parent.Name = i.ToString()
                parentBindingSource.Add(parent)
            Next
        End Sub

        Private Sub gridView1_MasterRowGetRelationCount(ByVal sender As Object, ByVal e As MasterRowGetRelationCountEventArgs)
            e.RelationCount = 1
        End Sub

        Private Sub gridView1_MasterRowEmpty(ByVal sender As Object, ByVal e As MasterRowEmptyEventArgs)
            e.IsEmpty = False
        End Sub

        Private Sub gridView1_MasterRowGetRelationName(ByVal sender As Object, ByVal e As MasterRowGetRelationNameEventArgs)
            e.RelationName = "Children"
        End Sub

        Private Sub gridView1_MasterRowGetChildList(ByVal sender As Object, ByVal e As MasterRowGetChildListEventArgs)
            Dim details As BindingList(Of Child) = New BindingList(Of Child)()
            e.ChildList = details
            Dim worker As BackgroundWorker = New BackgroundWorker()
            worker.DoWork += New DoWorkEventHandler(AddressOf worker_DoWork)
            worker.RunWorkerAsync(details)
        End Sub

        Private Sub worker_DoWork(ByVal sender As Object, ByVal e As DoWorkEventArgs)
            Dim details As BindingList(Of Child) = CType(e.Argument, BindingList(Of Child))
            For i As Integer = 0 To 100 - 1
                BeginInvoke(New MethodInvoker(Function()
                    Dim child As Child = New Child()
                    child.Name = i.ToString()
                    details.Add(child)
                End Function))
                Thread.Sleep(1000)
            Next
        End Sub
    End Class

    Public Class Parent

        Private _Name As String

        Public Property Name As String
            Get
                Return _Name
            End Get

            Set(ByVal value As String)
                _Name = value
            End Set
        End Property
    End Class

    Public Class Child

        Private _Name As String

        Public Property Name As String
            Get
                Return _Name
            End Get

            Set(ByVal value As String)
                _Name = value
            End Set
        End Property
    End Class
End Namespace

File: CS/WindowsFormsSample/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.Xpo;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Base;
using DevExpress.Data.Filtering;
using DevExpress.XtraGrid.Columns;
using System.Reflection;
using System.Threading;

namespace WindowsFormsSample {
    public partial class Form1 : Form {
        public Form1() {
            InitializeComponent();
        }
        private void Form1_Load(object sender, EventArgs e) {
            for (int i = 0; i < 5; i++) {
                Parent parent = new Parent();
                parent.Name = i.ToString();
                parentBindingSource.Add(parent);
            }
        }

        private void gridView1_MasterRowGetRelationCount(object sender, MasterRowGetRelationCountEventArgs e) {
            e.RelationCount = 1;
        }

        private void gridView1_MasterRowEmpty(object sender, MasterRowEmptyEventArgs e) {
            e.IsEmpty = false;
        }

        private void gridView1_MasterRowGetRelationName(object sender, MasterRowGetRelationNameEventArgs e) {
            e.RelationName = "Children";
        }

        private void gridView1_MasterRowGetChildList(object sender, MasterRowGetChildListEventArgs e) {
            BindingList<Child> details = new BindingList<Child>();
            e.ChildList = details;
            BackgroundWorker worker = new BackgroundWorker();
            worker.DoWork += new DoWorkEventHandler(worker_DoWork);
            worker.RunWorkerAsync(details);
        }

        void worker_DoWork(object sender, DoWorkEventArgs e) {
            BindingList<Child> details = (BindingList<Child>)e.Argument;
            for (int i = 0; i < 100; i++) {
                BeginInvoke(new MethodInvoker(delegate() {
                    Child child = new Child();
                    child.Name = i.ToString();
                    details.Add(child);
                }));
                Thread.Sleep(1000);
            }
        }
    }
    public class Parent {
        private string _Name;
        public string Name {
            get { return _Name; }
            set {
                _Name = value;
            }
        }
    }
    public class Child {
        private string _Name;
        public string Name {
            get { return _Name; }
            set {
                _Name = value;
            }
        }
    }
}
